### PR TITLE
Reduce recurring steps in CI workflow

### PR DESCRIPTION
**Pull request by nineinchnick**
## Description

Extracted from #12817

## Non-technical explanation

n/a

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (MiguelWeezardo) [inline .github/workflows/ci.yml:384]: This [action](https://github.com/trinodb/github-actions/blob/main/checkout-and-setup/action.yml) has `input.cache`default set to `maven` which is not the same as the [default](https://github.com/actions/setup-java/blob/de1bb2b0c5634f0fc4438d7aa9944e68f9bf86cc/action.yml#L55) from [setup-java](https://github.com/actions/setup-java/tree/v3) and means it's not a simple drop-in replacement as it changes behavior in some jobs. This job is a good example since it assumes there would be no attempt to update the cache with whatever `$RETRY $MAVEN validate` downloads, and makes no attempt to delete the local Maven repo.

Maybe the composite action should remove the whole Maven repo at the end, and only leave it behind if a special input is set? This input would only be set in `maven-checks` job. That way most jobs would not need any action inputs.

**Review by MiguelWeezardo — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml
Comment (nineinchnick) [reply .github/workflows/ci.yml:384]: Thanks, I need to adjust inputs here, this commit is older than the recent work we did around the cache. The action sets the cache to `maven` by default since this is the most common input for most jobs, but we do need to disable it here.

Unfortunately, a composite action cannot register post-job hooks.

Comment (nineinchnick) [reply .github/workflows/ci.yml:386]: I explicitly disabled cache in `web-ui-checks` and here in `build-test-matrix` jobs.

File: .github/workflows/ci.yml
Comment (hashhar) [inline .github/workflows/ci.yml:52]: must live inside the repo instead of being versioned outside - easy to test and easy to see what is actually being done.

Comment (hashhar) [reply .github/workflows/ci.yml:52]: the checkout is a one-liner anyway so can be done before the composite action and composite action can do other heavy lifting.

Comment (nineinchnick) [reply .github/workflows/ci.yml:52]: It's not a one-liner:
```
      - uses: actions/checkout@v3
        with:
          fetch-depth: 0 # checkout all commits, as the build result depends on `git describe` equivalent
```
Also, for #12817 we will need to verify if the checked-out commit is the one that's allowed, so we'll need an external action anyway. I'd rather avoid the additional work of creating a local action here and then removing it. I have an impression that we want #12817 to get merged soon, but I thought we could get these changes merged even sooner, as it simplifies the workflow and makes it easier to maintain, especially in forks.

Note that #12817 is open for 5 months, I keep rebasing it on a regular basis but never had to update the action (it exists for 4 months). I don't expect it to change that often, so I'm not sure there's a lot to gain if we'd split it up just to keep one definition in this repo.

**Comment by MiguelWeezardo** [comment]
Shouldn't the caching from #14882 be part of the composite action used here?

**Comment by nineinchnick** [comment]
> Shouldn't the caching from #14882 be part of the composite action used here?

Yes, that's why I opened https://github.com/trinodb/github-actions/pull/15 and I'll update the commit sha here when necessary.

**Review by MiguelWeezardo — APPROVED** [review]
Make sure to update the `checkout-and-setup` tag to something which uses `actions/cache` explicitly.

File: .github/workflows/ci.yml
Comment (nineinchnick) [reply .github/workflows/ci.yml]: I moved the action here and left the checkout step.

**Comment by nineinchnick** [comment]
I also extracted uploading test results and reports into a separate action and added a step to grab heap dumps.

**Comment by findepi** [comment]
@nineinchnick 
who can access the heap dumps? where are they stored and for how long?

**Comment by nineinchnick** [comment]
Ah, I forgot that they can contain sensitive data, like secrets. If we upload them as artifacts, then they're public. I'll add a condition to only do this if there are no secrets set.

**Comment by findepi** [comment]
This will work for now. If we have OOMs with secrets-requiring jobs, we can perhaps encrypt those dumps.

File: .github/actions/upload/action.yml
Comment (findepi) [inline .github/actions/upload/action.yml:9]: Can we afford 7 here? or 14?
2 is pretty low value

(i know that heap dumps are big)

File: .github/workflows/ci.yml
Comment (findepi) [inline .github/workflows/ci.yml:192]: make conditional?

this job has no secrets, but what if someone adds secrets later?
they may easily miss the fact we upload heap dumps

File: .github/workflows/ci.yml
Comment (findepi) [inline .github/workflows/ci.yml:303]: Let's try to simplify.

Can we somehow explicitly detect the PR is not trusted?
If github doesn't make it explicit, we could check that the build is a PR and the source repo != trinodb/trino

wdyt?

(i am worried that addition of a new secret will be overlook here)

alternatively, we could have a dummy secret SECRETS_PRESENT in trinodb/trino
and condition on that here

Comment (nineinchnick) [reply .github/workflows/ci.yml:303]: > alternatively, we could have a dummy secret SECRETS_PRESENT in trinodb/trino
and condition on that here

I like this idea. I didn't try to add a condition to detect forks because I'm not sure we wouldn't miss some cases. Checking for secrets seemed more direct.

**Comment by nineinchnick** [comment]
> @nineinchnick would be possible to decouple? (seems not obvious, but maybe...)

Decoupling would mean I'd have to do this twice, once by duplicating it 4 times and then again in this PR. Would it help if I'd extract it into another PR based on this branch?

File: .github/actions/upload/action.yml
Comment (nineinchnick) [reply .github/actions/upload/action.yml:9]: I think a higher value should be fine since this doesn't happen that often. I'll update this to 14.

File: .github/workflows/ci.yml
Comment (nineinchnick) [reply .github/workflows/ci.yml:303]: Can you create such a secret? Or would it be enough to check for just one of the existing secrets, with a comment?

**Comment by findepi** [comment]
> Decoupling would mean I'd have to do this twice, once by duplicating it 4 times and then again in this PR.

sounds bad
how bad?

> Would it help if I'd extract it into another PR based on this branch?

no, because i still wouldn't be able to merge this :)

File: .github/actions/setup/action.yml
Comment (hashhar) [inline .github/actions/setup/action.yml:9]: should this just be a boolean instead

File: .github/workflows/ci.yml
Comment (findepi) [reply .github/workflows/ci.yml:303]: I created org-wide SECRETS_PRESENT secret.

File: .github/actions/upload/action.yml
Comment (hashhar) [inline .github/actions/upload/action.yml:11]: stringly typed booleans create confusion specially when the value being used can be easily confused for a boolean

File: .github/workflows/ci.yml
Comment (hashhar) [reply .github/workflows/ci.yml:303]: why is checking for PR source not enough? i.e. source repo != trinodb/trino?

**Review by hashhar — COMMENTED** [review]
upload-heap-dump should be very carefully used - I'm not strongly in favour of enabling it by default. We should enable it when debugging for a specific job and not always.

Also are the heap dump artifacts only visible to people with write access to repo? If not then it's not safe to do at all.

**Comment by hashhar** [comment]
Let's extract last 2 commits and merge the first one. the last 2 are controversial.

**Comment by nineinchnick** [comment]
> Let's extract last 2 commits and merge the first one. the last 2 are controversial.

Done, I'll open up another PR after this one gets merged.

File: .github/actions/setup/action.yml
Comment (nineinchnick) [reply .github/actions/setup/action.yml]: Done

**Comment by hashhar** [comment]
>  > Also are the heap dump artifacts only visible to people with write access to repo? If not then it's not safe to do at all.

> Why they're not safe? Do you have ideas for any alternatives?

I now see that you only upload them if secrets don't exist so that solves my concern.

File: .github/actions/setup/action.yml
Comment (hashhar) [inline .github/actions/setup/action.yml:8]:
```suggestion
    description: "Cache Maven repo"
```

**Review by hashhar — APPROVED** [review]
LGTM.

File: .github/actions/upload/action.yml
Comment (MiguelWeezardo) [inline .github/actions/upload/action.yml:2]: There's no heap dump uploaded here aymore

**Review by MiguelWeezardo — APPROVED** [review]

**Review by hashhar — APPROVED** [review]
LGTM % `check-commits` can also use `.github/actions/setup` it seems.

**Comment by nineinchnick** [comment]
@hashhar all green!

**Review by hashhar — APPROVED** [review]

**Pull request merged**
